azafea/event_processors/endless/metrics/events/__init__.py

```diff
@@ -784,12 +793,22 @@ def _get_fields_from_payload(payload: GLib.Variant) -> Dict[str, Any]:
 
 # -- Sequence events ----------------------------------------------------------
 
+def default_time_duration(context):
+    parameters = context.get_current_parameters()
+    return (parameters['stopped_at'] - parameters['started_at']).total_seconds()
+
+
 class ShellAppIsOpen(SequenceEvent):
     __tablename__ = 'shell_app_is_open'
     __event_uuid__ = 'b5e11a3d-13f8-4219-84fd-c9ba0bf3d1f0'
     __payload_type__ = 's'
 
     app_id = Column(Unicode, nullable=False)
+    duration = Column(Float, default=default_time_duration, nullable=False)
+
+    __table_args__ = (
+        Index('ix_shell_app_is_open_app_id_started_at', 'started_at', 'app_id'),
```

Comment:
Is it possible to make the `app_id` field use `varchar_pattern_ops`? Based on https://www.postgresql.org/docs/9.4/indexes-opclass.html I believe this may be necessary to make queries like the following hit the index:

```sql
select * from shell_app_is_open
where started_at > now() - interval '30 days'
and app_id LIKE 'com.endlessm.encyclopedia.%'
```

though I would be equally happy to learn that the default operator class is enough to make postgres understand it can use the index for that prefix match.

Reply:
> Is it possible to make the `app_id` field use `varchar_pattern_ops`?

You’re right, that would be useful.

> though I would be equally happy to learn that the default operator class is enough to make postgres understand it can use the index for that prefix match.

I’ve tried on dev data, and it looks like we have to explicitly set the option. Creating the index is slower but we get much faster requests. Hopefully SQLAlchemy provides a simple syntax for that.

Reply:
I’ve added `varchar_pattern_ops`.